[processing] port some algorithms to C++

alexbruy opened the pull request:
## Description
Port several QGIS alorithms to C++:
- Extract specific vertices
- Geometry by expression
- Nearest neighbour analysis + add test for it
- Points (pixel centroids) from polygons
- Pole of inaccessibility
- Raster layer statistics
- Set raster style and Set vector style was replaced with single generic algorithm Set layer style. Old algorithms still here to keep backward compatibility but hidden from toolbox
- Truncate table
- Zonal statistics + add helper methods to `QgsZonalStatistics` to retrieve user-friendly statistics name from enum value
- Random extract. Also fixed unreported issue which caused incomplete results when generated random numbers list contains duplicated values. In such case old implementation outputs feature only once producing less features than requested.

## Checklist
- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

alexbruy committed:
[processing] port add table field algorithm to C++

alexbruy committed:
[processing] port sum line length algorithm to C++

alexbruy committed:
[processing] port extract layer extent to C++

alexbruy committed:
rename processing qgis test to allow running only subset of tests

alexbruy committed:
[processing] port extract specific vertices algorithm to C++

alexbruy committed:
[processing] port geometry by expression to C++

alexbruy committed:
[processing] port nearest neighbour analysis algorithm to C++ and add
test

alexbruy committed:
[processing] port points from polygons algorithm to C++

alexbruy committed:
[processing] port pole of inaccessibility to C++

alexbruy committed:
[processing] port raster layer statistics to C++

alexbruy committed:
[processing][needs-docs] replace set raster style and set vector style
Python algorithms with generics set layer style C++ algorithm

alexbruy committed:
[processing] port truncate table to C++

alexbruy committed:
[processing] port zonal statistics to C++

alexbruy committed:
[processing] port random extract to C++

Also fix unreported bug caused incomplete results when same FID
was used multiple times

root676 commented:
@alexbruy this is **huge**! Wow!

nyalldawson commented on python/plugins/processing/algs/qgis/SetRasterStyle.py:
```diff
@@ -48,7 +48,7 @@ def __init__(self):
 
     def flags(self):
-        return super().flags() | QgsProcessingAlgorithm.FlagNoThreading
+        return super().flags() | QgsProcessingAlgorithm.FlagNoThreading | QgsProcessingAlgorithm.FlagHideFromToolbox
```
The explicit "FlagDeprecated" would be more appropriate here (it also hides from the modeler, and would allow us to easily identify deprecated algorithms in future...)

alexbruy replied:
Done

nyalldawson commented on src/analysis/processing/qgsalgorithmextractspecificvertices.h:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Native extract specific vertices algorithm.
+ */
+class QgsExtractSpecificVerticesAlgorithm : public QgsProcessingAlgorithm
```
Hm, I wonder why we never made this one a feature based algorithm. Probably just an oversight. (We can address this later anyway)

alexbruy replied:
I was thinking about making it feature based but then looked at the "Extract vertices" C++ algorithm and found that itis also layer based. So I assumed there was some reasons to make it layer based and ported it as layer based too.

Will address this later if you don't object.

nyalldawson replied:
I'm guessing it was just an oversight -- I can't see any reason this couldn't/shouldn't be feature based.

nyalldawson commented on src/analysis/processing/qgsalgorithmnearestneighbouranalysis.cpp:
```diff
@@ -0,0 +1,155 @@
+QString QgsNearestNeighbourAnalysisAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm performs nearest neighbor analysis for a point layer.\n\n"
+                      "Output is generated as an html file with the computed statistical values." );
```
```suggestion
                      "Output is generated as an HTML file with the computed statistical values." );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmnearestneighbouranalysis.cpp:
```diff
@@ -0,0 +1,155 @@
+      break;
+    }
+
+    QgsFeatureId neighbourId = spatialIndex.nearestNeighbor( f.geometry().asPoint(), 2 ).at( 1 );
```
Hm, we can potentially give this algorithm a huge speed boost -- if we use the FlagStoreFeatureGeometries during the construction of the spatial index, then the geometries will be stored inside the algorithm and can be later retrieved using QgsSpatialIndex::geometry( ... ) . This would avoid the need for all the extra single-feature requests.

alexbruy replied:
Should be fixed now

nyalldawson commented on src/analysis/processing/qgsalgorithmpoleofinaccessibility.h:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Native pole of inaccessibility algorithm.
+ */
+class QgsPoleOfInaccessibilityAlgorithm : public QgsProcessingAlgorithm
```
Here's another one we should have made a feature based alg.

nyalldawson commented on src/analysis/processing/qgsalgorithmrandomextract.cpp:
```diff
@@ -0,0 +1,136 @@
+      break;
+    }
+
+    source->getFeatures( QgsFeatureRequest( id ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks ).nextFeature( f );
```
One by one feature requests are always quite slow. Looks like the original alg looped through the whole layer to avoid this. So while the new code will be much faster for low selection counts, when the count get close to the feature count it'll be much slower. 

I think you could avoid this by creating a single feature request with a .setFilterFids on it, and iterating through that. Note that you'd need to count the number of times that feature appears in the fids list and add that many duplicate features though (or you'll reintroduce the original bug!)

On that note, I think it might be worthwhile exposing a enum choice for sampling behaviour -- i.e. "sample with replacement" or "sample without replacement".

alexbruy replied:
My initial idea was exactly the same - use fids filter to query only required features. Unfortunately this way we loose duplicate features if no additional logic will be added. So I decided to go for slower option for now. Will look at it again and see if I can make it better.

roya0045 replied:
Out of curiosity, would there be a way to do one request for the base set and another round for duplicates by combining the fids with an expression?

alexbruy replied:
@roya0045 have you read comments above? This will be addressed in the follow-up pull-request

roya0045 replied:
@alexbruy yes but I was just trying to suggest a way to do things eventually.

alexbruy replied:
this was already suggested above

> you could avoid this by creating a single feature request with a .setFilterFids on it, and iterating through that. Note that you'd need to count the number of times that feature appears in the fids list and add that many duplicate features though (or you'll reintroduce the original bug!)

roya0045 replied:
This was not quite what I had in mind, I was talking about a first phase using the fids, and then for the second pass use fids + an expression to obtain the duplicates that you need instead of doing the exact same query twice or multiple times.

By having those two lists you would not have to care about reading things multiple times.

But this hinges on the fact that there is an optimal way to do this by using an expression and I am not familiar enough with the alg to know if this is true, hence why I was asking.

nyalldawson commented on src/analysis/processing/qgsalgorithmsumlinelength.cpp:
```diff
@@ -0,0 +1,214 @@
+    std::unique_ptr< QgsGeometryEngine > engine( QgsGeometry::createGeometryEngine( polyGeom.constGet() ) );
+    engine->prepareGeometry();
+
+    QgsDistanceArea da;
```
I'd move this up to the prepare step, or use a unique_ptr member so that we only need to initialize once (it's not cheap to create these)

alexbruy replied:
Done

nyalldawson commented on src/analysis/processing/qgsalgorithmzonalstatistics.cpp:
```diff
@@ -0,0 +1,153 @@
+
+QgsProcessingAlgorithm::Flags QgsZonalStatisticsAlgorithm::flags() const
+{
+  return QgsProcessingAlgorithm::flags() | QgsProcessingAlgorithm::FlagNoThreading;
```
I don't suggest doing it in this PR -- but what do you think about us deprecating this alg, and adding a new one which doesn't modify layers in place? I think it's more in line with how processing works...

alexbruy replied:
Sounds like a plan.

alexbruy replied:
Do you think we should also update `QgsZonalStatistics` class to produce new layer?

nirvn replied:
Yes please.

nyalldawson reviewed (commented):
Epic effort -- lovely work!

There's also some updates which will be required in gui/menus.py to swap qgis -> native (this doesn't happen automatically there)

alexbruy committed:
fix strings

alexbruy committed:
use deprecated flag instead of hidden

alexbruy committed:
move QgsDistanceArea initialization to prepareAlgorithm

alexbruy committed:
const correctness

alexbruy committed:
optimize nearest neighbour analysis by storing geometries in the spatial
index and using index to retrieve them instead of making single-feature
requests

alexbruy commented on src/analysis/processing/qgsalgorithmextractspecificvertices.cpp:
```diff
@@ -0,0 +1,197 @@
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  QString verticesString = parameterAsString( parameters, QStringLiteral( "VERTICES" ), context );
+  QStringList verticesList = verticesString.split( ',', QString::SkipEmptyParts );
```
Fixed

alexbruy committed:
move guts of the set layer style to prepareAlgorithm() to keep it
threaded

alexbruy commented on src/analysis/processing/qgsalgorithmapplylayerstyle.cpp:
```diff
@@ -0,0 +1,90 @@
+
+QgsProcessingAlgorithm::Flags QgsApplyLayerStyleAlgorithm::flags() const
+{
+  return QgsProcessingAlgorithm::flags() | QgsProcessingAlgorithm::FlagNoThreading;
```
Should be fixed. Hope I did it correctly.

alexbruy committed:
update menu entries

alexbruy committed:
mark unused parameters

alexbruy committed:
strore layer id instead of pointer to layer
